### PR TITLE
feature/PI-493-generating_party_keys

### PR DESCRIPTION
**Pull request by megan-bower4**

### CODE REVIEW

File: src/layers/domain/core/device_key/v3.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+        self.ods_code = ods_code
+        super().__init__(key_name="PARTYKEYNUMBER")
+
+    def _format_key(self) -> str:
+        """
+        Format the party key with the ODS code and a 7-digit number.
+        """
+        return f"{self.ods_code}-{self.current_number:07d}"
+
+    def validate_key(self, key: str) -> bool:
```
Comment (jameslinnell) [inline src/layers/domain/core/device_key/v3.py:91]: Is the validate key method needed? Surely _format_key() already makes sure of this?

File: src/layers/domain/core/device_key/tests/test_device_key_v3.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+
+def test_key_generator_instantiation():
+    # Instantiating KeyGenerator directly should raise TypeError
+    with pytest.raises(TypeError):
+        KeyGenerator(last_used_number=100)
+
+
+def test_party_key_generator_format_key():
+    generator = PartyKeyGenerator(last_used_number=123, ods_code="ABC")
```
Comment (jameslinnell) [inline src/layers/domain/core/device_key/tests/test_device_key_v3.py:15]: do these tests still work? last_used_number, is not an argument anymore.

File: src/layers/domain/core/device_key/v3.py (hunk 1 above)
Comment (megan-bower4) [reply src/layers/domain/core/device_key/v3.py:91]: Good point! As discussed it's useful for when we come to implementing the validation of provided/existing party_keys :)

File: src/layers/domain/core/device_key/tests/test_device_key_v3.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+from domain.core.device_key.v3 import KeyGenerator, PartyKeyGenerator
+from event.aws.client import dynamodb_client
+
+from test_helpers.terraform import read_terraform_output
+
+
+def test_key_generator_instantiation():
+    # Instantiating KeyGenerator directly should raise TypeError
+    with pytest.raises(TypeError):
+        KeyGenerator(last_used_number=100)
```
Comment (jameslinnell) [inline src/layers/domain/core/device_key/tests/test_device_key_v3.py:11]: KeyGenerator(key_name="PARTYKEYNUMBER")

**Review by jameslinnell — APPROVED** [review]

**Pull request closed without merging**